data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/ServerlessNetworkPolicyUpdater.java

```diff
@@ -0,0 +1,164 @@
+import software.amazon.awssdk.services.opensearchserverless.model.CreateSecurityPolicyRequest;
+import software.amazon.awssdk.services.opensearchserverless.model.GetSecurityPolicyRequest;
+import software.amazon.awssdk.services.opensearchserverless.model.GetSecurityPolicyResponse;
+import software.amazon.awssdk.services.opensearchserverless.model.ResourceNotFoundException;
+import software.amazon.awssdk.services.opensearchserverless.model.SecurityPolicyDetail;
+import software.amazon.awssdk.services.opensearchserverless.model.SecurityPolicyType;
+import software.amazon.awssdk.services.opensearchserverless.model.UpdateSecurityPolicyRequest;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+public class ServerlessNetworkPolicyUpdater {
+
+    static final String COLLECTION = "collection";
+    static final String DESCRIPTION = "Description";
+    static final String RESOURCE = "Resource";
+    static final String RESOURCE_TYPE = "ResourceType";
+    static final String RULES = "Rules";
+    static final String SOURCE_VPCES = "SourceVPCEs";
+
+    private static final Logger LOG = LoggerFactory.getLogger(ServerlessNetworkPolicyUpdater.class);
+
+    private final OpenSearchServerlessClient client;
+
+    public ServerlessNetworkPolicyUpdater(OpenSearchServerlessClient client) {
+        this.client = client;
+    }
+
+    public void updateNetworkPolicy(
+        final String networkPolicyName,
+        final String collectionName,
+        final String vpceId
+    ) {
+        try {
+            final Document newStatement = createNetworkPolicyStatement(collectionName, vpceId);
+            final Optional<SecurityPolicyDetail> maybeNetworkPolicy = getNetworkPolicy(networkPolicyName);
+
+            if (maybeNetworkPolicy.isPresent()) {
+                final Document existingPolicy = maybeNetworkPolicy.get().policy();
+                final String policyVersion = maybeNetworkPolicy.get().policyVersion();
+                final List<Document> existingStatements = existingPolicy.asList();
+                if (hasAcceptablePolicy(existingStatements, collectionName, vpceId)) {
+                    LOG.info("Policy statement already exists that matches collection and vpce id");
+                    return;
+                }
+
+                final List<Document> statements = new ArrayList<>(existingStatements);
+                statements.add(newStatement);
+                final Document newPolicy = Document.fromList(statements);
+                updateNetworkPolicy(networkPolicyName, newPolicy, policyVersion);
+            } else {
+                final Document newPolicy = Document.fromList(List.of(newStatement));
+                createNetworkPolicy(networkPolicyName, newPolicy);
+            }
+        } catch (final Exception e) {
+            LOG.error("Failed to create or update network policy", e);
+        }
+    }
+
+    // VisibleForTesting
```

Comment:
This class is relying a lot on testing isolated methods rather than the desired behavior. We can improve the tests by either: 1) testing the correct behavior; or 2) perhaps splitting this class into multiple class based on the individual behaviors. Taking the second route would allow you to mock behaviors from certain classes in your testing.

Reply:
Again, there are tests that test all the branches in the public method. The tests on the helper methods are more for my sanity than anything else. I can remove the visibility and delete the tests on the helper methods if you prefer.